### PR TITLE
Jetty http client async support

### PR DESCRIPTION
**Pull request by meltsufin**
Re-creating the PR from a local branch.
I added a small fix to the pom to make the image include jetty-client.jar.
@olamy

### CODE REVIEW

File: appengine-managed-runtime/src/main/java/com/google/apphosting/vmruntime/VmApiProxyDelegate.java [hunk 1]
```diff
@@ -98,29 +97,46 @@
...
+  public VmApiProxyDelegate() {
+    this(createHttpClient());
   }
 
-  public VmApiProxyDelegate() {
-    this(new DefaultHttpClient(createConnectionManager()));
+  private static HttpClient createHttpClient() {
+    try {
+      HttpClientTransportOverHTTP transport = new HttpClientTransportOverHTTP();
+      // FIXME what sort of customize SslContextFactory we want to use here??
```
Comment (meltsufin) [inline appengine-managed-runtime/src/main/java/com/google/apphosting/vmruntime/VmApiProxyDelegate.java]: It doesn't look like any requests go over SSL. So, we should be ok with not using one at all.

File: appengine-managed-runtime/src/main/java/com/google/apphosting/vmruntime/VmApiProxyDelegate.java [hunk 2]
```diff
@@ -98,29 +97,46 @@
...
+      SslContextFactory sslContextFactory = new SslContextFactory();
+      HttpClient httpClient = new HttpClient( transport, sslContextFactory );
+      httpClient.setMaxConnectionsPerDestination( VmApiProxyEnvironment.MAX_CONCURRENT_API_CALLS );
+      httpClient.setConnectTimeout( ADDITIONAL_HTTP_TIMEOUT_BUFFER_MS );
```
Comment (meltsufin) [inline appengine-managed-runtime/src/main/java/com/google/apphosting/vmruntime/VmApiProxyDelegate.java]: Originally it was `timeoutMs + ADDITIONAL_HTTP_TIMEOUT_BUFFER_MS`. I know that `timeoutMs` is not available at this point, but maybe there is a workaround?

File: appengine-managed-runtime/src/main/java/com/google/apphosting/vmruntime/VmApiProxyDelegate.java [hunk 3]
```diff
@@ -360,48 +381,34 @@ RuntimeException constructApiException(String packageName, String methodName) {
...
-        CoreConnectionPNames.SO_TIMEOUT, timeoutMs + ADDITIONAL_HTTP_TIMEOUT_BUFFER_MS);
+    request.header(RPC_STUB_ID_HEADER, REQUEST_STUB_ID);
+    request.header(RPC_METHOD_HEADER, REQUEST_STUB_METHOD);
 
-    // Performance tweaks.
-    params.setBooleanParameter(CoreConnectionPNames.TCP_NODELAY, Boolean.TRUE);
-    params.setBooleanParameter(CoreConnectionPNames.STALE_CONNECTION_CHECK, Boolean.FALSE);
-    request.setParams(params);
+
+    request.timeout( ADDITIONAL_HTTP_TIMEOUT_BUFFER_MS, TimeUnit.MILLISECONDS );
```
Comment (meltsufin) [inline appengine-managed-runtime/src/main/java/com/google/apphosting/vmruntime/VmApiProxyDelegate.java]: missing `timeoutMs +`?

File: appengine-managed-runtime/src/main/java/com/google/apphosting/vmruntime/VmApiProxyDelegate.java (hunk 1 above)
Comment (olamy) [inline appengine-managed-runtime/src/main/java/com/google/apphosting/vmruntime/VmApiProxyDelegate.java]: correct we only have to remove the FIXME :-)

File: appengine-managed-runtime/src/main/java/com/google/apphosting/vmruntime/VmApiProxyDelegate.java (hunk 2 above)
Comment (olamy) [inline appengine-managed-runtime/src/main/java/com/google/apphosting/vmruntime/VmApiProxyDelegate.java]: @meltsufin AFAICS the connection timeout was alway set to DEFAULT_RPC_TIMEOUT_MS. So I will add to have similar behaviour.

**Comment by meltsufin** [comment]
LGTM

**Pull request merged**